imu_integrator/plugin.cpp

```diff
@@ -0,0 +1,477 @@
+		// 	cam_count++;
+		// 	last_cam_time = input_values->last_cam_integration_time;
+		// 	std::cout << "Num IMUs recieved since last cam: " << counter << " Diff between new cam and latest IMU: " 
+		// 			  << timestamp - last_cam_time << " Expected IMUs recieved VS Actual: " << cam_count*10 << ", " << total_imu << std::endl;
+		// 	counter = 0;
+		// }
+		// counter++;
+
+		// Get what our IMU-camera offset should be (t_imu = t_cam + calib_dt)
+		double t_off_new = input_values->t_offset;
+
+		// This is the last CAM time
+		double time0 = input_values->last_cam_integration_time + last_imu_offset;
+		double time1 = timestamp + t_off_new;
+
+		vector<ov_msckf::Propagator::IMUDATA> prop_data = select_imu_readings(_imu_vec, time0, time1);
+		ov_type::IMU *temp_imu = new ov_type::IMU();
+		temp_imu->set_value(input_values->imu_value);
+		temp_imu->set_fej(input_values->imu_fej);
+
+		Eigen::Matrix<double,3,1> w_hat;
+		Eigen::Matrix<double,3,1> a_hat;
+		Eigen::Matrix<double,3,1> w_hat2;
+		Eigen::Matrix<double,3,1> a_hat2;
+
+		// Loop through all IMU messages, and use them to move the state forward in time
+		// This uses the zero'th order quat, and then constant acceleration discrete
+		if (prop_data.size() > 1) {
+			for(size_t i=0; i<prop_data.size()-1; i++) {
+
+				// Time elapsed over interval
+				double dt = prop_data.at(i+1).timestamp-prop_data.at(i).timestamp;
+
+				// Corrected imu measurements
+				w_hat = prop_data.at(i).wm - temp_imu->bias_g();
+				a_hat = prop_data.at(i).am - temp_imu->bias_a();
+				w_hat2 = prop_data.at(i+1).wm - temp_imu->bias_g();
+				a_hat2 = prop_data.at(i+1).am - temp_imu->bias_a();
+
+				// Compute the new state mean value
+				Eigen::Vector4d new_q;
+				Eigen::Vector3d new_v, new_p;
+				predict_mean_rk4(temp_imu->quat(), temp_imu->pos(), temp_imu->vel(), input_values->gravity,
+								dt, w_hat, a_hat, w_hat2, a_hat2, new_q, new_v, new_p);
+
+				//Now replace imu estimate and fej with propagated values
+				Eigen::Matrix<double,16,1> imu_x = temp_imu->value();
+				imu_x.block(0,0,4,1) = new_q;
+				imu_x.block(4,0,3,1) = new_p;
+				imu_x.block(7,0,3,1) = new_v;
+				temp_imu->set_value(imu_x);
+				temp_imu->set_fej(imu_x);
+			}
+		}
+
+		Eigen::Matrix<double,13,1> state_plus = Eigen::Matrix<double,13,1>::Zero();
+		state_plus.block(0,0,4,1) = temp_imu->quat();
+		state_plus.block(4,0,3,1) = temp_imu->pos();
+		state_plus.block(7,0,3,1) = temp_imu->vel();
+
+		if (prop_data.size() > 1) state_plus.block(10,0,3,1) = prop_data.at(prop_data.size()-2).wm - temp_imu->bias_g();
+		else if (!prop_data.empty()) state_plus.block(10,0,3,1) = prop_data.at(prop_data.size()-1).wm - temp_imu->bias_g();
+
+		_m_imu_raw->put(new imu_raw_type{
+			w_hat,
+			a_hat,
+			w_hat2,
+			a_hat2,
+			state_plus,
+		});
+    }
+
+	std::vector<ov_msckf::Propagator::IMUDATA> select_imu_readings(const std::vector<ov_msckf::Propagator::IMUDATA>& imu_data, double time0, double time1) {
```

Comment:
If this came from OpenVINS, drop a comment saying so (and a pointer to the original source).